GVFS/GVFS/CommandLine/UpgradeVerb.cs

```diff
@@ -0,0 +1,309 @@
+                return false;
+            }
+
+            return true;
+        }
+
+        private bool TryRunUpgradeChecks(
+            out Version latestVersion,
+            out bool installable,
+            out string error)
+        {
+            bool upgradeCheckSuccess = false;
+            bool upgradeInstallable = false;
+            string errorMessage = null;
+            Version version = null;
+
+            this.ShowStatusWhileRunning(
+                () =>
+                {
+                    upgradeCheckSuccess = this.TryCheckUpgradeAvailable(out version, out errorMessage);
+                    if (upgradeCheckSuccess && version != null)
+                    {
+                        upgradeInstallable = true;
+
+                        if (!this.TryCheckUpgradeInstallable(out errorMessage))
+                        {
+                            upgradeInstallable = false;
+                        }
+                    }
+
+                    return upgradeCheckSuccess;
+                },
+                 "Checking for GVFS upgrades",
+                suppressGvfsLogMessage: true);
+
+            latestVersion = version;
+            installable = upgradeInstallable;
+            error = errorMessage;
+
+            return upgradeCheckSuccess;
+        }
+
+        private bool TryRunInstaller(out string error)
+        {
+            string upgraderPath = null;
+            string errorMessage = null;
+
+            bool preUpgradeSuccess = this.ShowStatusWhileRunning(
+                () =>
+                {
+                    if (this.TryCopyUpgradeTool(out upgraderPath, out errorMessage) &&
+                        this.TryLaunchUpgradeTool(upgraderPath, out errorMessage))
+                    {
+                        return true;
+                    }
+
+                    return false;
+                },
+                "Launching upgrade tool",
+                suppressGvfsLogMessage: true);
+            
+            if (!preUpgradeSuccess)
+            {
+                error = errorMessage;
+                return false;
+            }
+
+            error = null;
+            return true;
+        }
+
+        private bool TryCopyUpgradeTool(out string upgraderExePath, out string error)
+        {
+            upgraderExePath = null;
+
+            this.tracer.RelatedInfo("Copying upgrade tool");
+
+            if (!this.upgrader.TryCreateToolsDirectory(out upgraderExePath, out error))
+            {
+                return false;
+            }
+
+            this.tracer.RelatedInfo("Successfully Copied upgrade tool.");
+
+            return true;
+        }
+
+        private bool TryLaunchUpgradeTool(string path, out string error)
+        {
+            this.tracer.RelatedInfo("Launching upgrade tool");
+
+            if (!this.processWrapper.Start(path))
+            {
+                error = "Error launching upgrade tool";
+                return false;

```

Comment:
Log a warning\error here?  Also, would it be possible to log other details about what failed?

Reply:
Absolutely. Updated to capture Process.Start exceptions when available.